Enable Ninja Builds in Visual Studio

DannyHindson opened the pull request:
**Description of work.**

Allow a ninja build including use of ccache to be run from within Visual Studio 2019 or 2022 using the CMake integration features. The main change to allow this was increasing the version number in the CMakePresets.json file from 1 to 2 which then means Visual Studio parses it. I've made some other related changes:

- clarified the use of MSVC in the dev docs. Both the ninja and Visual Studio generators use the MSVC compiler on Windows so I've changed the name of the win-msvc preset to win-vs
- changed the dev docs to suggest new starters install Visual Studio 2022 instead of 2019. Both are still supported but seems sensible now to switch to the newer version
- increased the minimum required cmake version in the CMakePresets.json file to 3.21 (which is the version where the VS 2022 generator is available)
- Removed the logic that only allocated targets to folders for the Visual Studio generator. Folders specified when using the Ninja generator are read when the resulting project is opened in Visual Studio
- fixed a problem where running with CONDA_ENV=false followed by CONDA_ENV=true in the same source directory caused an error during the cmake configure step. I've set the Python_EXECUTABLE cmake variable explicitly in the CONDA_ENV=true logic. This switch seems more likely to happen when using the CMake targets view in Visual Studio where the default preset selected can be something other than "win-ninja"

The general experience in Visual Studio is approximately similar to what you get when using the Visual Studio cmake generator. The main difference is that you can't start up workbench by setting the "workbench" target as a start up item. So I've been running workbench from PyCharm instead. I think whatever configuration allows this to work with the Visual Studio generator isn't available with the ninja set-up. Also you can't right click on a folder (eg MantidFramework) and build all the targets inside the folder. I've seen online that this is available in a preview version of Visual Studio but it's not in my version of VS2022 (which is v17.4.2)

**To test:**

Probably worth cloning mantid into a new source code directory in case this doesn't work for any reason and trashes your main directory: `git clone https://github.com/mantidproject/mantid.git mantidninja`

Follow the dev-docs instructions in:

a) the CCache parts of BuildingWithCMake.html
b) and the updated GettingStartedCondaWindows.html page - run the ninja cmake option using `cmake --preset=win-ninja -B build` and then choose the "Compile and Build using Ninja" option

Open up the Ninja configured build in Visual Studio with the solution explorer set to "CMake Targets View" and the Configuration set to "win-ninja". The target system\configuration\build dropdowns should look like:

![image](https://user-images.githubusercontent.com/56295817/207627982-381d9a79-7221-4d4e-9f70-e0e608ba0aef.png)

...and the solution explorer should look like:

![image](https://user-images.githubusercontent.com/56295817/207627110-4b716c77-89e6-44ff-be36-1351d6ff8880.png)

1) Select Build, Build All. This will build everything except the tests
2) Once this completes check the incremental build is working OK by repeating and observing that no work is done
3) Right click on the "Mantid Project" entry in the tree and select "Delete cache and Reconfigure" 
4) Select Build, Build All. This should use CCache to rebuild the code more quickly than 1)

*There is no associated issue.*

*This does not require release notes* because **there's no functional change to end users, only affects developers**

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Enable compilation using Ninja in Visual Studio

Visual Studio requires the schema to be v2 (even though the rest of the content
is valid in the v1 preset json schema). v2 is only available in CMake 3.20 so
move cmakeMinimumRequired on in here also. My interpretation of cmakeMinimumRequired
is that it's the minimum version required to make sense of the presets json file
so not syncing this up with cmake_minimum_required in the main CMakeLists.txt

Add cmakeExecutable setting to ninja preset. This stops VS using it's internal
CMake executable and instead causes VS to look for a cmake.exe on the path - which
will be the CMake version in Conda if VS has been started up from an active conda
environment

Also rename the win-msvc preset to be win-vs since both the windows presets
use MSVC as the underlying compiler

DannyHindson committed:
Add preset for VS 2022

Make this the default with the VS2019 preset having the suffix. Seems
mature enough now to get new starters using VS2022
Generator only available in cmake 3.21 so move on min required

DannyHindson committed:
Few tidy ups

Make Visual Studio use the ninja version in Conda rather than its own version
Fix problem where you run cmake twice changing CONDA_ENV from False to True by
setting the Python_EXECUTABLE variable (more likely to happen if running cmake
in Visual Studio where the initial preset may not be the one you want on opening
Visual Studio)
Run cmake commands to add targets to folders when running ninja generator. These
folders still get used if you then open up the ninja generated source folders in
Visual Studio

DannyHindson committed:
Stop cache refresh being required every time start up Visual Studio

DannyHindson committed:
Add batch file to start up VS with ninja generated build config

DannyHindson committed:
Update documentation on how to use ninja generator in VS

DannyHindson commented:
I found an issue with Visual Studio which is that the cmake cache refreshes on start up even if nothing has changed. This isn't the end of the world but it takes a few seconds to regenerate the cache. I submitted this issue to Microsoft and it's currently "Under Consideration":

https://developercommunity.visualstudio.com/t/CMake-cache-requires-refresh-on-starting/10232092

robertapplin commented on dev-docs/source/GettingStarted/GettingStartedCondaWindows.rst:
```diff
@@ -53,34 +53,42 @@ Setup the mantid conda environment
+    * Open Visual Studio using ``visual-studio_ninja.bat``
+    * Select "win-ninja" from the Configuration dropdown at the top of the screen
+    * Select "Switch between solutions and available views" in the Solution Explorer and click on CMake Targets View
+    * It is recommmended that the option "When cache is out of date" option is set to "Never run configure step automatically" in Tools, Options, CMake
```
```suggestion
    * It is recommmended that the "When cache is out of date" option is set to "Never run configure step automatically" in Tools->Options->CMake
```

robertapplin commented on dev-docs/source/GettingStarted/GettingStartedCondaWindows.rst:
```diff
@@ -53,34 +53,42 @@ Setup the mantid conda environment
+    * To build the unit tests use: ``ninja AllTests``
+
+* In Visual Studio:
+    * Open Visual Studio using ``visual-studio_ninja.bat``
```
```suggestion
    * Open Visual Studio with ``visual-studio_ninja.bat``, which is found in the build folder
```

robertapplin commented on dev-docs/source/GettingStarted/GettingStartedCondaWindows.rst:
```diff
@@ -53,34 +53,42 @@ Setup the mantid conda environment
+* In Visual Studio:
+    * Open Visual Studio using ``visual-studio_ninja.bat``
+    * Select "win-ninja" from the Configuration dropdown at the top of the screen
+    * Select "Switch between solutions and available views" in the Solution Explorer and click on CMake Targets View
```
```suggestion
    * Select "Switch between solutions and available views" in the Solution Explorer and click on "CMake Targets View"
```

robertapplin commented on dev-docs/source/GettingStarted/GettingStartedCondaWindows.rst:
```diff
@@ -53,34 +53,42 @@ Setup the mantid conda environment
+----------------------------------------
 
-* Open visual studio with ``visualstudio.bat`` then click build.
+* Open visual studio with ``visual-studio.bat`` then click build.
```
```suggestion
* Open visual studio with ``visual-studio.bat``, which is found in the build folder, and then click build.
```

robertapplin reviewed (changes requested):
I'm testing this at the moment and its all looking good - its a big improvement for Ninja users on Windows! I just had a few documentation suggestions that I thought I'd send in the meantime

robertapplin commented:
Something I noticed after doing `Build`->`Build All` is the files are always built into the <source_dir>/build folder even when I have specified the build folder to be outside of the source. So maybe the build directory is not being picked up properly when opening visual studio?

I have tried to change the `buildRoot` variable in the `CMakeSettings.json` file that seems to be generated (when opening visual studio?), but this didn't seem to change fix the build directory problem either

robertapplin commented:
In the top right of visual studio it has an option to `Select startup Item`. I assume we will be able to select "bin/workbench.exe" and this will allow us to open Mantid? If so it might be worth adding an instruction for this too

DannyHindson commented:
> In the top right of visual studio it has an option to `Select startup Item`. I assume we will be able to select "bin/workbench.exe" and this will allow us to open Mantid? If so it might be worth adding an instruction for this too

If you look in the tree view on the right hand side (CMake Targets View) you'll see that Visual Studio classifies each target as shared library or utility target or custom command. As far as I can see you can only select targets of type "shared library" as start up items. The workbench target is classed as a utility target so you can't select it. I think the problem is more fundamental than just not having the option available in the Visual Studio UI - I think the workbench target only works because of the ability of the Visual Studio cmake generator to take on board a start up command using the VS_DEBUGGER_COMMAND setting and this just isn't supported by the ninja generator. So think we'll need to just fire up workbench from PyCharm

DannyHindson commented:
> Something I noticed after doing `Build`->`Build All` is the files are always built into the <source_dir>/build folder even when I have specified the build folder to be outside of the source. So maybe the build directory is not being picked up properly when opening visual studio?
> 
> I have tried to change the `buildRoot` variable in the `CMakeSettings.json` file that seems to be generated (when opening visual studio?), but this didn't seem to fix the build directory problem

Just trying out changing the build path settings. So if I do something more gentle such as changing binaryDir on the win-ninja preset in CMakePresets.json from `${sourceDir}/build` to `${sourceDir}/buildxxxx` this seems to work. Are you saying that if I change it to `c:\build` so it's completely outside my source code folder it doesn't work?

robertapplin commented:
> Just trying out changing the build path settings. So if I do something more gentle such as changing binaryDir on the win-ninja preset in CMakePresets.json from `${sourceDir}/build` to `${sourceDir}/buildxxxx` this seems to work. Are you saying that if I change it to `c:\build` so it's completely outside my source code folder it doesn't work?

Ah ok, I haven't tried changing the build directory inside the `CMakePresets.json` file. Instead I specified a different build directory from the command line:
```
cmake --preset=win-ninja -B=../buildninja
```

Does this work for you?

DannyHindson committed:
Use actual source code root rather than assuming ..

DannyHindson committed:
Minor review comment updates

DannyHindson commented:
Just noticed one other small difference between the CMake Targets View and the original "solution" view in Visual Studio - there doesn't seem to be a cancel build option in the CMake Targets View so you just have to wait for it to finish

DannyHindson commented:
> Just noticed one other small difference between the CMake Targets View and the original "solution" view in Visual Studio - there doesn't seem to be a cancel build option in the CMake Targets View so you just have to wait for it to finish

Actually you can - I was building a single target by right clicking on the target and selecting Build Algorithms (for example). Although there's no cancel option on the main Build menu there is one that's appeared on the right click menu option of the target you're building

robertapplin reviewed (dismissed):
Thanks for the changes, and also for looking into those other points I raised. This looks good to go 👍

martyngigg commented on CMakePresets.json:
```diff
@@ -52,16 +52,30 @@
+      "displayName": "win-vs",
+      "description": "Default build options for a windows visual studio conda build",
+      "binaryDir": "${sourceDir}/build",
+      "generator": "Visual Studio 17 2022",
```
I think at the moment this doesn't hit us as we don't use super advance features in C++ but Conda are using VS2019 (https://conda-forge.org/docs/user/announcements.html#moving-to-visual-studio-toolchain-vc142) and are therefore using the `vc141` toolset. For 2022 can we also set the toolset version to `vc141` so that we can use the IDE but match the toolset for condaforge to ensure compatability: https://cmake.org/cmake/help/latest/variable/MSVC_TOOLSET_VERSION.html

martyngigg replied:
Perhaps we should then also get developers to move to 2022 as a default?

DannyHindson replied:
As discussed I'll try and force the toolset version in cmake. I guess I should force to `vc142` rather than `vc141`?

Re getting developers to move to VS2022 - yes that's my intention. I've updated the "Getting Started with Windows..." doc page to say developers should install 2022. I've been less definite on using the VS generator vs ninja. The instructions just explain how to set up either.

DannyHindson replied:
I've just pushed the changes to force the v142 toolset. A few comments on this:
- I've put the change in the CMakePresets.json file because it's convenient but also because I had to - the CMake documentation suggests setting the toolset in your CMake project code is unsupported. I did manage to get it to work that way by setting the CMake variable right at the start before the initial `project(Mantid)` line in the main CMakeLists.txt file but it seems a bit risky and also a bit odd putting it there
- the v142 toolset isn't installed with VS2022 by default so I've added a note to the documentation about that
 - the change only affects the Visual Studio generator. The ninja generator doesn't support the toolset version being specified so still some risk of v143 being used there
- the Visual Studio 2022 UI highlights the fact we're running the old toolset in quite a verbose way but not a big deal I think:

![image](https://user-images.githubusercontent.com/56295817/217585936-ba811544-c6e2-498e-9969-d061c05ac498.png)

DannyHindson replied:
OK thanks. We're getting the compiler tools into the `PATH` at the moment for ninja users by advising developers to run the `cmake` commands from the Visual Studio command prompt. The Visual Studio 2022 command prompt has the v143 toolset in its path though so that's the problem we'd have to fix

DannyHindson committed:
Use v142 toolset with Visual Studio 2022

This is to align with Conda which currently uses v142. Do this
to reduce risk of any incompatibility betweeen Mantid and Conda
dependencies
Ninja generator doesn't support toolset specification so still
some risk that could use more recent toolset version
Have specified in the preset partly because adding logic to the
CMake project code involves setting the toolset before the
initial "project" call at the top of the root CMakeLists.txt
